src/vm/prestub.cpp

```diff
@@ -250,90 +243,346 @@ void DACNotifyCompilationFinished(MethodDesc *methodDesc)
+
+    if (IsILStub())
+    {
+        ILStubResolver* pResolver = AsDynamicMethodDesc()->GetILStubResolver();
+        return pResolver->GetILHeader();
     }
     else
     {
-        fStable = FALSE;
-        flags.Add(CORJIT_FLAGS(CORJIT_FLAGS::CORJIT_FLAG_TIER0));
+        return NULL;
     }
-#endif
 
+    // NoMetadata currently doesn't verify the IL. I'm not sure if that was
+    // a deliberate decision in the past or not, but I've left the behavior
+    // as-is during refactoring.
+}
+
+COR_ILMETHOD_DECODER* MethodDesc::GetAndVerifyILHeader(PrepareCodeConfig* pConfig, NewHolder<COR_ILMETHOD_DECODER> & holder)
+{
+    STANDARD_VM_CONTRACT;
+    _ASSERTE(IsIL() || IsNoMetadata());
+
+    if (IsNoMetadata())
+    {
+        // The NoMetadata version doesn't allocate the decoder in response 
+        // to this call so it doesn't need the holder
+        return GetAndVerifyNoMetadataILHeader();
+    }
+    else
+    {
+        return GetAndVerifyMetadataILHeader(pConfig, holder);
+    }
+}
+
+// ********************************************************************
+//                  README!!
+// ********************************************************************
+
+// JitCompileCode is the thread safe way to invoke the JIT compiler
+// If multiple threads get in here for the same config, ALL of them
+// MUST return the SAME value for pcode.
+//
+// This function creates a DeadlockAware list of methods being jitted
+// which prevents us from trying to JIT the same method more that once.
+
+PCODE MethodDesc::JitCompileCode(PrepareCodeConfig* pConfig)
+{
+    STANDARD_VM_CONTRACT;
+
+    LOG((LF_JIT, LL_INFO1000000,
+        "JitCompileCode(" FMT_ADDR ", %s) for %s:%s\n",
+        DBG_ADDR(this),
+        IsILStub() ? " TRUE" : "FALSE",
+        GetMethodTable()->GetDebugClassName(),
+        m_pszDebugMethodName));
+
+    PCODE pCode = NULL;
     {
         // Enter the global lock which protects the list of all functions being JITd
-        ListLockHolder pJitLock (GetDomain()->GetJitLock());
+        JitListLock::LockHolder pJitLock(GetDomain()->GetJitLock());
 
         // It is possible that another thread stepped in before we entered the global lock for the first time.
-        pCode = GetNativeCode();
-        if (pCode != NULL)
+        if (pConfig->IsJitCancellationRequested(&pCode))
         {
-#ifdef FEATURE_INTERPRETER
-            if (Interpreter::InterpretationStubToMethodInfo(pCode) == this)
-            {
-                pPreviousInterpStub = pCode;
-            }
-            else
-#endif // FEATURE_INTERPRETER
-            goto Done;
+            return pCode;
         }
 
         const char *description = "jit lock";
         INDEBUG(description = m_pszDebugMethodName;)
-        ListLockEntryHolder pEntry(ListLockEntry::Find(pJitLock, this, description));
+            ReleaseHolder<JitListLockEntry> pEntry(JitListLockEntry::Find(
+                pJitLock, pConfig->GetCodeVersion(), description));
```

Comment:
Because this entry is now version specific, is it possible that we have a race between two threads attempting to compile the same IL but with different target versions (e.g. different optimization levels), and if so, is it possible that a higher optimization level might lose out to a lower opt-level?